src/test/py/bazel/bzlmod/bazel_lockfile_test.py

```diff
@@ -1899,6 +1897,70 @@ def testExtensionRepoMappingChange_mainRepoEvalCycleWithWorkspace(self):
     # Build again. This should _NOT_ trigger a failure!
     _, _, stderr = self.RunBazel(['build', '--enable_workspace', ':lol'])
     self.assertNotIn('ran the extension!', '\n'.join(stderr))
+  def testExtensionIgnoredInLockfile(self):
+    self.ScratchFile(
+      'MODULE.bazel',
+      [
+        'ext1 = use_extension("extension.bzl", "ext1")',
+        'ext2 = use_extension("extension.bzl", "ext2")',
+        'use_repo(ext1, "repo1")',
+        'use_repo(ext2, "repo2")',
+      ],
+    )
+    self.ScratchFile('BUILD.bazel')
+    self.ScratchFile(
+      'extension.bzl',
+      [
+        'def _repo_rule_impl(ctx):',
+        '    ctx.file("BUILD", "filegroup(name=\'lala\')")',
+        'repo_rule = repository_rule(implementation=_repo_rule_impl)',
+        '',
+        'def _should_lock_impl(ctx): repo_rule(name="repo1")',
+        'def _no_lock_impl(ctx):',
+        '    repo_rule(name="repo2")',
+        '    return ctx.extension_metadata(',
+        '        root_module_direct_deps=[],',
+        '        root_module_direct_dev_deps=[],',
+        '        exclude_from_lockfile=True',
+        '    )',
+        'ext1 = module_extension(implementation=_should_lock_impl)',
+        'ext2 = module_extension(implementation=_no_lock_impl)',
+      ],
+    )
+
+    self.RunBazel(['build', '@repo1//:all', '@repo2//:all'])
+    with open(self.Path('MODULE.bazel.lock'), 'r') as f:
+      lockfile = json.loads(f.read().strip())
+      self.assertIn('//:extension.bzl%ext1', lockfile['moduleExtensions'])
+      self.assertNotIn('//:extension.bzl%ext2', lockfile['moduleExtensions'])
+
+    # Update extensions implementations to the opposite
+    self.ScratchFile(
+      'extension.bzl',
+      [
+        'def _repo_rule_impl(ctx):',
+        '    ctx.file("BUILD", "filegroup(name=\'lala\')")',
+        'repo_rule = repository_rule(implementation=_repo_rule_impl)',
+        '',
+        'def _should_lock_impl(ctx): repo_rule(name="repo2")',
+        'def _no_lock_impl(ctx):',
+        '    repo_rule(name="repo1")',
+        '    return ctx.extension_metadata(',
+        '        root_module_direct_deps=[],',
+        '        root_module_direct_dev_deps=[],',
+        '        exclude_from_lockfile=True',
+        '    )',
+        'ext1 = module_extension(implementation=_no_lock_impl)',
+        'ext2 = module_extension(implementation=_should_lock_impl)',
+      ],
+    )
+
+    # Assert updates in the lockfile
+    self.RunBazel(['build', '@repo1//:all', '@repo2//:all'])
+    with open(self.Path('MODULE.bazel.lock'), 'r') as f:
+      lockfile = json.loads(f.read().strip())
+      self.assertNotIn('//:extension.bzl%ext1', lockfile['moduleExtensions'])
+      self.assertIn('//:extension.bzl%ext2', lockfile['moduleExtensions'])
```

Comment:
Could you add an invocation with `--lockfile_mode=error`? I'm not entirely sure, but I could see that resulting in an error since ext1 won't have an entry in the lockfile. But this behavior would make the `error` mode unusable.

Reply:
You're right. The lockfile expects the extension to be locked, and it fails. I guess the only way to ignore that check is to have this as part of the extension definition not metadata. A bit less flexible but better than losing the error mode. WDYT?

Reply:
alternatively, we could have a special sentinel value for this extension in the lockfile, signaling that the extension was run but the results are not stored here (and that it's safe to rerun the extension).

Reply:
My initial idea was to just unlocked extensions and let them fail when they don't set `reproducible = True`, but that can delay CI failures. I like the idea of storing a sentinel value.

To cover all use cases I have encountered so far, we would need to
* allow `reproducible` to be set dynamically: Some extensions such as r_j_e are reproducible if and only if the users specifies a domain-specific lockfile via a tag, which isn't statically known.
* if `reproducible` is set, store the sentinel value in a way that applies to every os/arch combination and document that reproducibility must not depend on the host platform: This is needed for extensions that generate toolchains (e g. go_sdk and the Python extensions), which are typically reproducible and platform-dependent. Being able to skip the lockfile completely avoids having to update it on each supported platform.

Reply:
Hmmm I see a few things to handle:
The problem is here "store the sentinel value in a way that applies to every os/arch combination", When we read from the lockfile, we don't know yet if the extension is reproducible or not. So we are searching using the usual eval factors we have. To pass that we have to:
- Override those factors with empty values when we write/ send event for the extension.
- Override the extension value with empty (why store any other data? except for the sentinel value)
- Search Twice when we read, with eval factors and with empty eval factors.
Better ideas?

I feel like we are complicating the logic here, and storing empty things that are not very useful in the lockfile. So my question is: Are the cons of having them in the metadata worth it?

Reply:
> Override the extension value with empty (why store any other data? except for the sentinel value)

I like this option best. Having an extension listed with an empty body doesn't feel like a hack to track the "this has run successfully once, but we didn't have to store any data about it" bit.

> Are the cons of having them in the metadata worth it?

I personally think that's the case: The current reality of module extensions is that many have their own lockfile format, either because they inherited it from WORKSPACE or because dependency resolution in that ecosystem is inherently complicated (Python) or non-reproducible (Rust, see Daniel's comments on the lockfile format doc). If we want extension authors to gradually migrate off of custom lockfile formats and on to `MODULE.bazel.lock` where possible, this "transition" has to be configurable via extension tags.

Reply:
The more I think about it, the more I like the alternative of just not storing any lockfile entry for an extension that is `reproducible`. Even storing an empty entry still requires users to run `bazel mod deps` once before the build passes with `--lockfile_mode=error` and, as @SalmaSamy remarked, clutters the lockfile compared to static `reproducible` declaration.

The only downside of not storing entries is that we would need to evaluate extensions without entries even with `--lockfile_mode=error` and fail if they don't return metadata indicating that they are reproducible. But that would only result in slightly delayed feedback in the case where a user forgets to trigger a newly added non-reproducible extension at least once before pushing to CI. 

What do you think of this alternative?

Reply:
Hmmm okay I see the benefits here, and I agree that I like 'not storing it at all' more!

I wasn't really thinking of options, I was thinking we will have to do all of these steps :D
If we only do the empty extensions thing, then:
- We will have multiple empty extensions in the lockfile (specially with various os and arch values for the ones that depend on them).
- We are spending time reading, parsing and writing stuff, when the lockfile should not be even used in this case
- If the reproducible value changes, what happens to the old entries? Previously I thought we should delete them but now? delete, keep but empty or just keep them (is it possible to set the reproducible tag depending on the os?)

So yeah, maybe just running it again is (sadly) the way to go. @Wyverald WDYT?